lib/puppetserver/ca/revoke_action.rb

```diff
@@ -0,0 +1,123 @@
+require 'optparse'
+require 'openssl'
+require 'net/https'
+require 'json'
+
+module Puppetserver
+  module Ca
+    class RevokeAction
+      # Actions require a short summary constant
+      SUMMARY = 'Revoke a given certificate'
+      BANNER = <<-BANNER
+Usage:
+  puppetserver ca revoke [--help|--version]
+  puppetserver ca revoke CERTNAME
+
+Description:
+Given a valid certname, instructs the CA to revoke it.
+
+Options:
+BANNER
+
+      # class method parser returns an instance of Ruby's OptionParser
+      # (or a custom object that implements summarize() and help().
+      def self.parser(parsed = {})
+        OptionParser.new do |o|
+          o.banner = BANNER
+          o.on('--help', 'Display this revoke specific help output') do |help|
+            parsed['help'] = true
+          end
+          o.on('--version', 'Output the version') do |v|
+            parsed['version'] = true
+          end
+        end
+      end
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      # take unparsed array of arguments, return parsed hash and optionally an error code
+      def parse(args)
+        results = { certname: [] }
+        unparsed = []
+        parser = self.class.parser(results)
+
+        begin
+          parser.order!(args) do |unknown|
+            results[:certname] << unknown
+          end
+        rescue OptionParser::ParseError => e
+          unparsed += e.args
+          unparsed << args.shift unless args.first =~ /^-{1,2}/
+          retry
+        end
+
+        exit_code = nil
+        unless unparsed.empty?
+          @logger.err 'Error:'
+          @logger.err 'Unknown arguments or flags:'
+          unparsed.each do |arg|
+            @logger.err "    #{arg}"
+          end
+
+          @logger.err ''
+          @logger.err parser.help
+          exit_code = 1
+        end
+
+        if results[:certname].length == 0
+          @logger.err 'A certname to revoke is required'
+          @logger.err parser.help
+          exit_code = 1
+        elsif results[:certname].length > 1
+          @logger.err 'Only one certname is allowed at a time'
+          @logger.err 'Received certnames:'
+          results[:certname].each do |certname|
+            @logger.err "    #{certname}"
+          end
+          exit_code = 1
+
+        else
+          results[:certname] = results[:certname].first
```

Comment:
Shouldn't reassign a hash value like this, should maybe store the args in an intermediate place that we do validation on before assigning certname to result.

Reply:
Yeah that seems better. I mostly don't like that the type changes.